### PR TITLE
Feature224 envvarforscripts

### PR DESCRIPTION
**Pull request by julienmalik**
Just tested on my prod server. It works.

Sample output of an installation.

```
admin@ynh:~/ynh/example_ynh$ sudo yunohost app remove ynhexample ; sudo yunohost app install . --debug
Success! ynhexample successfully removed
Success! SSOwat configuration successfully generated
149  DEBUG loading actions map namespace 'yunohost'
162  DEBUG extra parameter classes loaded: ['ask', 'password', 'required', 'pattern']
163  DEBUG initializing base actions map parser for cli
167  DEBUG registering new callback action 'yunohost.utils.packages.ynh_packages_version' to ['-v', '--version']
387  DEBUG initialize authenticator 'ldap-anonymous' with: uri='ldap://localhost:389', base_dn='dc=yunohost,dc=org', user_rdn='None'
391  DEBUG lock has been acquired
470  INFO processing action [16695.1]: yunohost.app.install with args={'app': '.', 'args': None, 'auth': <moulinette.authenticators.ldap.Authenticator object at 0x7fc311c28150>, 'label': None}
484  INFO Extracting...
503  INFO Done.
Choose a domain name for ynhexample: yolo.org
Choose a path for ynhexample (default: /example): 
Choose an admin user: yoloman
Is it a public application? [Yes | No] (default: Yes): No
Do you want to check this boolean ? [0 | 1] (default: 1): 
13402 INFO Executing command 'sudo -u admin -H sh -c cd "/var/cache/yunohost/from_file/scripts" && YNH_APP_INSTANCE_NUMBER="1" YNH_APP_ARG_my_boolean="1" YNH_APP_ARG_is_public="No" YNH_APP_ARG_admin="yoloman" YNH_APP_ID="ynhexample" YNH_APP_ARG_path="/example" YNH_APP_ARG_domain="yolo.org" BASH_XTRACEFD=7 /bin/bash -x "./install" "yolo.org" "/example" "yoloman" "No" "1" "ynhexample" 7>&1'...
13456 INFO + set -e
13459 INFO + for arg in domain path admin is_public my_boolean
13460 INFO + envvar=YNH_APP_ARG_domain
13461 INFO + echo 'YNH_APP_ARG_domain = yolo.org'
13462 INFO YNH_APP_ARG_domain = yolo.org
13463 INFO + for arg in domain path admin is_public my_boolean
13464 INFO + envvar=YNH_APP_ARG_path
13465 INFO + echo 'YNH_APP_ARG_path = /example'
13466 INFO YNH_APP_ARG_path = /example
13467 INFO + for arg in domain path admin is_public my_boolean
13468 INFO + envvar=YNH_APP_ARG_admin
13468 INFO + echo 'YNH_APP_ARG_admin = yoloman'
13469 INFO YNH_APP_ARG_admin = yoloman
13471 INFO + for arg in domain path admin is_public my_boolean
13471 INFO + envvar=YNH_APP_ARG_is_public
13472 INFO + echo 'YNH_APP_ARG_is_public = No'
13473 INFO YNH_APP_ARG_is_public = No
13474 INFO + for arg in domain path admin is_public my_boolean
13475 INFO + envvar=YNH_APP_ARG_my_boolean
13476 INFO + echo 'YNH_APP_ARG_my_boolean = 1'
13477 INFO YNH_APP_ARG_my_boolean = 1
13477 INFO + echo YNH_APP_ID=ynhexample
13478 INFO YNH_APP_ID=ynhexample
13479 INFO + echo YNH_APP_INSTANCE_NUMBER=1
13480 INFO YNH_APP_INSTANCE_NUMBER=1
13480 INFO + app=ynhexample
13481 INFO + domain=yolo.org
13482 INFO + path=/example
13483 INFO + admin=yoloman
13483 INFO + is_public=No
```

I have only added this snippet at the beginning of the install script of the example_ynh app :
```
for arg in domain path admin is_public my_boolean ; do
  envvar=YNH_APP_ARG_${arg}
  echo "$envvar = ${!envvar}"
done
echo YNH_APP_ID=$YNH_APP_ID
echo YNH_APP_INSTANCE_NUMBER=$YNH_APP_INSTANCE_NUMBER
```

### CODE REVIEW

File: src/yunohost/app.py [hunk 1]
```diff
@@ -375,9 +376,15 @@ def app_upgrade(auth, app=[], url=None, file=None):
...
         # Retrieve arguments list for upgrade script
         # TODO: Allow to specify arguments
-        args_list = _parse_args_from_manifest(manifest, 'upgrade', auth=auth)
+        args_odict = _parse_args_from_manifest(manifest, 'upgrade', auth=auth)
+        args_list = args_odict.values()
         args_list.append(app_id)
 
+        # Prepare env. var. to pass to script
+        env_dict = _make_environment_dict(args_odict)
+        env_dict["YNH_APP_INSTANCE_NUMBER"] = app_id.split('__')[-1]
```
Comment (jeromelebleu) [inline src/yunohost/app.py]: Is the value `1` for the first application?

File: src/yunohost/app.py [hunk 2]
```diff
@@ -494,7 +507,7 @@ def app_install(auth, app, label=None, args=None):
     os.system('cp %s/manifest.json %s' % (app_tmp_folder, app_setting_path))
     os.system('cp -R %s/scripts %s' % (app_tmp_folder, app_setting_path))
     try:
-        if hook_exec(app_tmp_folder + '/scripts/install', args_list) == 0:
+        if hook_exec(app_tmp_folder + '/scripts/install', args=args_list, env=env_dict) == 0:
```
Comment (jeromelebleu) [inline src/yunohost/app.py]: It seems that you forgot to pass the `env` argument for the `app_upgrade` action, as you do there for `app_install`.

File: src/yunohost/app.py [hunk 3]
```diff
@@ -1497,9 +1509,22 @@ def _parse_args_from_manifest(manifest, action, args={}, auth=None):
...
+    Convert a dictionnary containing manifest arguments
+    to a dictionnary of env. var. to be passed to scripts
+
+    Keyword arguments:
+        arg -- A key/value dictionnary of manifest arguments
+
+    """
+    env_dict = {}
+    for arg_name, arg_value in args_dict.items():
+        env_dict[ "YNH_APP_ARG_%s" % arg_name ] = arg_value
```
Comment (jeromelebleu) [inline src/yunohost/app.py]: I would prefer to format environment variables to be uppercase only in order to be more rigorous, as discussed on the bug tracker (I don't succeed to find the ticket anymore...).

**Comment by julienmalik** [comment]
@jeromelebleu I have addressed your two comments

File: src/yunohost/backup.py [hunk 4]
```diff
@@ -467,14 +476,22 @@ def _is_hook_in_backup(h):
...
-                          raise_on_error=True, chdir=tmp_app_bkp_dir)
+
+                # Prepare env. var. to pass to script
+                env_dict = {}
+                app_id, app_instance_nb = _parse_app_instance_name(app_instance_name)
+                env_dict["YNH_APP_ID"] = app_id
+                env_dict["YNH_APP_INSTANCE_NAME"] = app_instance_name
+                env_dict["YNH_APP_INSTANCE_NUMBER"] = str(app_instance_nb)
+
+                hook_exec(tmp_script, args=[tmp_app_dir + '/backup', app_instance_name],
```
Comment (jeromelebleu) [inline src/yunohost/backup.py]: `args` argument value has changed, it should be `args=[tmp_app_bkp_dir, app_instance_name],`.

**Comment by jeromelebleu** [comment]
It sounds good to me - except my minor comment - thanks! I'll just release a new testing version before actually testing and merging. Since there are lot of changes since the last one, it will be difficult to identify bugs otherwise...

File: src/yunohost/backup.py [hunk 5]
```diff
@@ -186,38 +188,47 @@ def _clean_tmp_dir(retcode=0):
...
                 subprocess.call(['install', '-Dm555', app_script, tmp_script])
-                hook_exec(tmp_script, args=[tmp_app_bkp_dir, app_id],
-                          raise_on_error=True, chdir=tmp_app_bkp_dir)
+
+                # Prepare env. var. to pass to script
+                env_dict = {}
+                app_id, app_instance_nb = _parse_app_instance_name(app_instance_name)
+                env_dict["YNH_APP_ID"] = app_id
+                env_dict["YNH_APP_INSTANCE_NAME"] = app_instance_name
+                env_dict["YNH_APP_INSTANCE_NUMBER"] = str(app_instance_nb)
```
Comment (jeromelebleu) [inline src/yunohost/backup.py]: The `tmp_app_bkp_dir` is not set in the environment variables, I think we should add it if we want to go to the very end!

File: src/yunohost/backup.py [hunk 6]
```diff
@@ -467,14 +478,22 @@ def _is_hook_in_backup(h):
...
```
Comment (jeromelebleu) [inline src/yunohost/backup.py]: And is not set here too...

File: src/yunohost/backup.py (hunk 5 above)
Comment (julienmalik) [inline src/yunohost/backup.py]: YNH_APP_BACKUP_DIR ok ?

Comment (jeromelebleu) [inline src/yunohost/backup.py]: Done.

**Comment by julienmalik** [comment]
Updated by #142

**Pull request closed without merging**